salt/pillar/nsot.py

```diff
@@ -0,0 +1,220 @@
+    query = salt.utils.http.query(url, data=data_dict, method='POST',
+                                  decode=True)
+    error = query.get('error')
+    if error:
+        log.error('Cannot obtain NSoT authentication token due to: %s.', error)
+        log.debug('Please verify NSoT URL %s is reachable and email %s is valid', url, email)
+        return False
+    else:
+        log.debug('successfully obtained token from nsot!')
+        return query['dict'].get('auth_token')
+
+
+def _check_regex(minion_id, regex):
+    '''
+    check whether or not this minion should have this external pillar returned
+
+    :param minion_id: str
+    :param minion_regex: list
+    :return: bool
+    '''
+    get_pillar = False
+    for pattern in regex:
+        log.debug('nsot external pillar comparing %s with %s', minion_id, regex)
+        match = re.search(pattern, minion_id)
+        if match and match.string == minion_id:
+            log.debug('nsot external pillar found a match!')
+            get_pillar = True
+            break
+        log.debug('nsot external pillar unable to find a match!')
+    return get_pillar
+
+
+def _query_nsot(url, headers, device=None):
+    '''
+    if a device is given, query nsot for that specific device, otherwise return
+    all devices
+
+    :param url: str
+    :param headers: dict
+    :param device: None or str
+    :return:
+    '''
+    url = urlparse.urljoin(url, 'devices')
+    if not device:
+        query = salt.utils.http.query(url, header_dict=headers, decode=True)
+    else:
+        url = urlparse.urljoin(url, device)
+        query = salt.utils.http.query(url, header_dict=headers,
+                                      decode=True)
+    error = query.get('error')
+    if error:
+        log.error('can\'t get device(s) from nsot! reason: %s', error)
+        return {}
+    else:
+        return query['dict']
+
+
+def _proxy_info(minion_id, api_url, email, secret_key, fqdn_separator):
+    '''
+    retrieve a dict of a device that exists in nsot
+
+    :param minion_id: str
+    :param api_url: str
+    :param email: str
+    :param secret_key: str
+    :param fqdn_separator: str
+    :return: dict
+    '''
+    if fqdn_separator:
+        minion_id = minion_id.replace('.', fqdn_separator)
+    token = _get_token(api_url, email, secret_key)
+    if not token:
+        return
+    headers = {'Authorization': 'AuthToken {}:{}'.format(email, token)}
+    device_info = _query_nsot(api_url, headers, device=minion_id)
+    return device_info
+
+
+def _all_nsot_devices(api_url, email, secret_key):
+    '''
+    retrieve a list of all devices that exist in nsot
+
+    :param api_url: str
+    :param email: str
+    :param secret_key: str
+    :return: dict
+    '''
+    token = _get_token(api_url, email, secret_key)
+    if not token:
+        return
+    headers = {'Authorization': 'AuthToken {}:{}'.format(email, token)}
+    all_devices = _query_nsot(api_url, headers)
+    return all_devices
```

Comment:
It would be better to return the same type. In both `_all_nsot_devices` and `_proxy_info` functions. Now you return `dictionary` or `None`. To fix this, please remove multuple returns:
```python
token = _get_token(api_url, email, secret_key)
if token:
    headers = {'Authorization': 'AuthToken {}:{}'.format(email, token)}
    all_devices = _query_nsot(api_url, headers)
else:
    all_devices = {}

return all_devices
```

Reply:
@wasabi222 Are you able to resolve this concern? If so, we should be able to merge this!

Reply:
@cachedout I just updated my PR to fix the last of these issues :)